pkg/flowcli/services/transactions.go

```diff
@@ -53,87 +50,159 @@ func NewTransactions(
 	}
 
-	return t.send(code, signer, args, argsJSON)
+	if waitSeal {
+		t.logger.StartProgress("Waiting for transaction to be sealed...")
+	}
+
+	result, err := t.gateway.GetTransactionResult(tx, waitSeal)
+
+	t.logger.StopProgress()
+
+	return tx, result, err
 }
 
-// SendForAddressWithCode send transaction for address and private key specified with code
-func (t *Transactions) SendForAddressWithCode(
-	code []byte,
-	signerAddress string,
-	signerPrivateKey string,
+// Sign transaction
+func (t *Transactions) Sign(
+	signerName string,
+	proposerName string,
+	payerAddress string,
+	additionalAuthorizers []string,
+	role string,
+	scriptFilename string,
+	payloadFilename string,
 	args []string,
 	argsJSON string,
-) (*flow.Transaction, *flow.TransactionResult, error) {
-	address := flow.HexToAddress(signerAddress)
+	approveSigning bool,
+) (*project.Transaction, error) {
+	if t.project == nil {
+		return nil, fmt.Errorf("missing configuration, initialize it: flow project init")
+	}
 
-	privateKey, err := crypto.DecodePrivateKeyHex(crypto.ECDSA_P256, signerPrivateKey)
-	if err != nil {
-		return nil, nil, fmt.Errorf("private key is not correct")
+	// get the signer account
+	signerAccount := t.project.AccountByName(signerName)
+	if signerAccount == nil {
+		return nil, fmt.Errorf("signer account: [%s] doesn't exists in configuration", signerName)
 	}
 
-	account := project.AccountFromAddressAndKey(address, privateKey)
+	if payloadFilename != "" && scriptFilename != "" {
+		return nil, fmt.Errorf("can not use both a transaction payload and Cadence code file")
+	} else if payloadFilename == "" && scriptFilename == "" {
+		return nil, fmt.Errorf("provide either a transaction payload or Cadence code file")
+	}
 
-	return t.send(code, account, args, argsJSON)
-}
+	// if we received already created transaction payload, create from payload and return signed
+	if payloadFilename != "" {
+		tx, err := project.NewTransactionFromPayload(signerAccount, payloadFilename, role)
+		if err != nil {
+			return nil, err
+		}
 
-func (t *Transactions) send(
-	code []byte,
-	signer *project.Account,
-	args []string,
-	argsJSON string,
-) (*flow.Transaction, *flow.TransactionResult, error) {
+		if approveSigning {
+			return tx.Sign()
+		}
 
-	// if google kms account then sign in TODO discuss refactor - move to account
-	if signer.DefaultKey().Type() == config.KeyTypeGoogleKMS {
-		resourceID := signer.DefaultKey().ToConfig().Context[config.KMSContextField]
-		err := util.GcloudApplicationSignin(resourceID)
-		if err != nil {
-			return nil, nil, err
+		if output.ApproveTransactionPrompt(tx) {
+			return tx.Sign()
+		} else {
+			return nil, fmt.Errorf("transaction was not approved for signing")
 		}
 	}
 
-	t.logger.StartProgress("Sending Transaction...")
+	// we are creating a new transaction
+	tx := project.NewTransaction()
 
-	tx := flow.NewTransaction().
-		SetScript(code).
-		AddAuthorizer(signer.Address())
+	err := tx.SetSigner(signerAccount)
```

Comment:
```suggestion
	err := tx.SetSigner(signerAccount, role)
```

Any reason not to do this? I'm not sure where else `SetSigner` is called, so apologies if this doesn't work.

Reply:
Yes it makes sense, I didn't implement it like this because it progressed to this solution and I missed it